### PR TITLE
Introduce job util

### PR DESCRIPTION
**Pull request by orlandohohmeier**
---
ℹ️   _This branch depends on features that will be introduced with #396 and #420, thus it will fail until the respective branches get merged._

---

This util is used to create job specs from job structs as well as to create from models from jobs and respective form schemas.

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @philipnrmn

File: src/js/utils/JobUtil.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import Job from '../structs/Job';
+
+const getFindPropertiesRecursive = function (job, item) {
```
Comment (MatApple) [inline src/js/utils/JobUtil.js]: Not sure about the naming of this :) maybe we could avoid the `getFind`?

Comment (orlandohohmeier) [inline src/js/utils/JobUtil.js]: I've to admit that I bluntly stole this one from the `ServiceUtil` and didn't thought about the naming. But yeah, let's find a better name for it.

File: src/js/utils/JobUtil.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    let spec = {};
+    let {general = {}, labels = {}, docker, schedule} = formModel;
+
+    spec.id = general.id;
+    spec.description = general.description;
+
+    if (Array.isArray(labels.items)) {
+      spec.labels = labels.items.reduce(function (memo, {key, value}) {
+        memo[key] = value || null;
+        return memo;
```
Comment (MatApple) [inline src/js/utils/JobUtil.js]: #nit lets have a newline above `return`

File: src/js/utils/JobUtil.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  createJobFromFormModel: function (formModel) {
+    if (formModel == null) {
+      return new Job();
+    }
+
+    let spec = {};
+    let {general = {}, labels = {}, docker, schedule} = formModel;
+
+    spec.id = general.id;
+    spec.description = general.description;
```
Comment (MatApple) [inline src/js/utils/JobUtil.js]: Should we have defaults for `general.id` and `general.description`?

Comment (orlandohohmeier) [inline src/js/utils/JobUtil.js]: Yeah, good idea. But let's do that in a separate PR.

File: src/js/utils/JobUtil.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      docker: docker
+    };
+
+    spec.schedules = [schedule];
+
+    return new Job(spec);
+  },
+
+  createFormModelFromSchema: function (schema, job = new Job()) {
+
```
Comment (aldipower) [inline src/js/utils/JobUtil.js]: remove extra line please

File: src/js/utils/JobUtil.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import Job from '../structs/Job';
+
+const getMatchingProperties = function (job, item) {
+
```
Comment (aldipower) [inline src/js/utils/JobUtil.js]: no extra line

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please

File: src/js/utils/JobUtil.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
+import Job from '../structs/Job';
+
+const getMatchingProperties = function (job, item) {
```
Comment (mlunoe) [inline src/js/utils/JobUtil.js]: Could we get a description of this function or some comments? Not sure what exactly is going on here.
Picking out properties from a schema, but what is the `getter`?

File: src/js/utils/__tests__/JobUtil-test.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        id: 'test',
+        cmd: 'sleep 1000;'
+      });
+
+      expect(JobUtil.createJobSpecFromJob(job)).toEqual(
+        {
+          id: 'test',
+          description: null,
+          run: {
+            cpus: 0.01,
```
Comment (mlunoe) [inline src/js/utils/__tests__/JobUtil-test.js]: Not sure I understand where these values come from?

File: src/js/utils/JobUtil.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        return memo;
+      }, {});
+    }
+
+    spec.run = {
+      cmd: general.cmd,
+      cpus: general.cpus,
+      mem: general.mem,
+      disk: general.disk,
+      docker: docker
```
Comment (mlunoe) [inline src/js/utils/JobUtil.js]: #nit this could be just `docker`

File: src/js/utils/__tests__/JobUtil-test.js (hunk 7 above)
Comment (orlandohohmeier) [inline src/js/utils/__tests__/JobUtil-test.js]: It's default as defined in https://github.com/dcos/dcos-ui/blob/master/src/js/structs/Job.js

File: src/js/utils/JobUtil.js (hunk 6 above)
Comment (Poltergeist) [inline src/js/utils/JobUtil.js]: This is getting the model from the schema. `getter` is a attribute on the schema which is an anonymous function which returns the value for that field from the struct. This is used because a) it gives in the schema the possibility to transform the data from the struct to the form. Think key value to array, or array to comma separated list. b) it is a fairly simple approach which you can see directly on the schema. So if a field is filled out wrong there is just one point to look at.

This function is actually walking recursively  trough the schema and creates the model for the form from a struct. The idea is to fill out dynamically the model by giving a schema and a struct in this case the `jobs` struct.

File: src/js/utils/JobUtil.js (hunk 3 above)
Comment (MatApple) [inline src/js/utils/JobUtil.js]: Couldn't we just put them in L28? `{general = {id: null, description: null}, `?

File: src/js/utils/JobUtil.js (hunk 6 above)
Comment (orlandohohmeier) [inline src/js/utils/JobUtil.js]: We should introduced this method in the `SchemaUtil` instead and is should be capable to simply  create modal from a schema and additionally augmenting it with data object like the structs in this example.  
As we don't have time at the moment, I'd like to focus on the actual feature and do this refactoring in the next dev cycle. What do you think?

File: src/js/utils/JobUtil.js (hunk 3 above)
Comment (orlandohohmeier) [inline src/js/utils/JobUtil.js]: Yeah, why not.

File: src/js/utils/JobUtil.js (hunk 6 above)
Comment (mlunoe) [inline src/js/utils/JobUtil.js]: Yes, let's create a TODO for it! An move on!

**Comment by mlunoe** [comment]
Feel free to merge once tests pass

**Comment by orlandohohmeier** [comment]
**mesosphere-ci** retest this please
